src/CarbonAware/src/Interfaces/ICarbonIntensityDataSource.cs

```diff
@@ -25,9 +25,10 @@ public interface ICarbonIntensityDataSource
     public Task<IEnumerable<EmissionsData>> GetCarbonIntensityAsync(IEnumerable<Location> locations, DateTimeOffset periodStartTime, DateTimeOffset periodEndTime);
 
     /// <summary>
-    /// Gets the forecasted carbon intensity for a location
+    /// Gets the forecasted carbon intensity for a location. If the optional `generatedAt` field is not provided, defaults to the current forecast.
     /// </summary>
     /// <param name="location">The location that should be used for getting the forecast.</param>
-    /// <returns>A forecasted emissions object for the given location.</returns>
-    public Task<EmissionsForecast> GetCurrentCarbonIntensityForecastAsync(Location location);
+    /// <param name="generatedAt">[Optional] The historical time used to fetch the most recent forecast generated as of that time.</param>
+    /// <returns>A forecasted emissions object for the given location generated at the given time.</returns>
+    public Task<EmissionsForecast> GetCarbonIntensityForecastAsync(Location location, DateTimeOffset? generatedAt = null);
```

Comment:
When implementing the historical forecast, we realized that the implementation almost 1:1 repeated the code used for the current forecast with only a handful of deviations. In order to keep the code cleaner and clearer for future use, we decided to unify the implementations at the datasource level and have a single function that handles both current and historical. If we want the current forecast, we only pass a location. If we want a historical forecast, we pass in a location and a generatedAt time. 

The aggregator takes care of passing the right combo of parameters based on whether it's doing the current or historical

Reply:
For WattTime this is true.  I wonder if this is true for the domain as a whole, or if it would vary between emissions API services.   Did you consider as an alternative, leaving both functions in the interface and dealing with the duplication in the implementation?  We do this in the [WattTimeClient](https://github.com/microsoft/carbon-aware-sdk/blob/dev/src/CarbonAware.Tools/CarbonAware.Tools.WattTimeClient/src/WattTimeClient.cs#L129-L132).

As a nice side-effect it would remove the need for optional null type checking throughout.

Reply:
Yeah I guess it could just be cleaner to handle separately, will do